Foursquare provider

rjperes opened the pull request.

rjperes committed:
Foursquare provider

AlexArchive commented on Owin.Security.Providers/Foursquare/FoursquareAuthenticationHandler.cs:
```diff
@@ -0,0 +1,238 @@
+					return new AuthenticationTicket(null, properties);
+				}
+
+				var graphResponse = await this._httpClient.GetAsync(GraphApiEndpoint + "?oauth_token=" + Uri.EscapeDataString(accessToken), Request.CallCancelled);
```
Hm. On my computer running on `localhost` this line results in a 410 "Gone". This causes the subsequent statement (`EnsureSuccessStatusCode`) to throw an exception. I assume this does not happen on your machine...?

rjperes replied:
No, it doesn't, but I never ran it in localhost... I use 127.0.0.1.xip.io and configure IIS Express accordingly. Can you try it with a non-local address, please?

rjperes committed:
- Added v=version with the current date in YYYYMMdd format;
- Added m=foursquare so as to use the Foursquare format (https://developer.foursquare.com/overview/versioning);
- Formatted code so as to match guidelines;
- Added claims to authenticated user identity;
- Added reference and commented code to demo project.

rjperes committed:
Fixed formatting

rjperes committed:
Fixed formatting

AlexArchive commented on Owin.Security.Providers/Foursquare/FoursquareAuthenticationHandler.cs:
```diff
@@ -0,0 +1,240 @@
+						new Claim("urn:foursquare:id", context.Id, XmlSchemaString, this.Options.AuthenticationType),
+						new Claim("urn:foursquare:name", context.Name, XmlSchemaString, this.Options.AuthenticationType),
+						new Claim("urn:foursquare:email", context.Email, XmlSchemaString, this.Options.AuthenticationType),
+						new Claim("urn:foursquare:twitter", context.Twitter, XmlSchemaString, this.Options.AuthenticationType)
```
It is possible that the authenticated user does not have a Twitter account or public email, and so on. You need to surround **every** claim in an `if` statement like [this](https://github.com/owin-middleware/OwinOAuthProviders/blob/master/Owin.Security.Providers/GitHub/GitHubAuthenticationHandler.cs#L98). If you do not, an exception will be thrown for users who are "missing" values.

rjperes replied:
Right

rjperes committed:
- Fixed formatting issues;
- Protected email claims;

AlexArchive commented on Owin.Security.Providers/Foursquare/FoursquareAuthenticationHandler.cs:
```diff
@@ -0,0 +1,243 @@
+				this.GenerateCorrelationId(extra);
+
+				// OAuth2 3.3 space separated
+				var scope = string.Join(" ", this.Options.Scope);
```
This variable is not assigned. If it is not needed can you please remove it and the `Scope` property from the options class. If it is needed, please update the query string. Thank you.

AlexArchive commented on Owin.Security.Providers/Foursquare/FoursquareAuthenticationHandler.cs:
```diff
@@ -0,0 +1,243 @@
+						"&state=" + Uri.EscapeDataString(state);
+
+				this.Response.StatusCode = (int)HttpStatusCode.Moved;
+				this.Response.Headers.Set("Location", authorizationEndpoint);
```
I think you can replace lines 188 and 189 with 

``` C#
Response.Redirect(authorizationEndpoint);
```

(No need to worry about this one if you do not want to.)

rjperes committed:
- Added scope query string parameter;
- Using Response.Redirect instead of headers + status code;
